pkg/instancetype/instancetype.go

```diff
@@ -333,6 +351,66 @@ func getInstancetypeAPISpec(obj runtime.Object) (interface{}, error) {
 	}
 }
 
+func checkCPUPreferenceRequirements(instancetypeSpec *instancetypev1beta1.VirtualMachineInstancetypeSpec, preferenceSpec *instancetypev1beta1.VirtualMachinePreferenceSpec, vmiSpec *virtv1.VirtualMachineInstanceSpec) (*k8sfield.Path, error) {
+	if instancetypeSpec != nil {
+		if instancetypeSpec.CPU.Guest < preferenceSpec.Requirements.CPU.Guest {
+			return k8sfield.NewPath("spec", "instancetype"), fmt.Errorf(InsufficientInstanceTypeCPUResourcesErrorFmt, instancetypeSpec.CPU.Guest, preferenceSpec.Requirements.CPU.Guest)
+		}
+	} else {
+		if vmiSpec.Domain.CPU == nil {
+			return k8sfield.NewPath("spec", "template", "spec", "domain", "cpu"), fmt.Errorf(NoVMCPUResourcesDefinedErrorFmt, preferenceSpec.Requirements.CPU.Guest)
+		}
+
+		switch getPreferredTopology(preferenceSpec) {
+		case instancetypev1beta1.PreferThreads:
+			if vmiSpec.Domain.CPU.Threads < preferenceSpec.Requirements.CPU.Guest {
+				return k8sfield.NewPath("spec", "template", "spec", "domain", "cpu", "threads"), fmt.Errorf(InsufficientVMCPUResourcesErrorFmt, vmiSpec.Domain.CPU.Threads, preferenceSpec.Requirements.CPU.Guest, "threads")
+			}
+		case instancetypev1beta1.PreferCores:
+			if vmiSpec.Domain.CPU.Cores < preferenceSpec.Requirements.CPU.Guest {
+				return k8sfield.NewPath("spec", "template", "spec", "domain", "cpu", "cores"), fmt.Errorf(InsufficientVMCPUResourcesErrorFmt, vmiSpec.Domain.CPU.Cores, preferenceSpec.Requirements.CPU.Guest, "cores")
+			}
+		case instancetypev1beta1.PreferSockets:
+			if vmiSpec.Domain.CPU.Sockets < preferenceSpec.Requirements.CPU.Guest {
+				return k8sfield.NewPath("spec", "template", "spec", "domain", "cpu", "sockets"), fmt.Errorf(InsufficientVMCPUResourcesErrorFmt, vmiSpec.Domain.CPU.Sockets, preferenceSpec.Requirements.CPU.Guest, "sockets")
+			}
+		}
+	}
+
+	return nil, nil
+}
+
+func checkMemoryPreferenceRequirements(instancetypeSpec *instancetypev1beta1.VirtualMachineInstancetypeSpec, preferenceSpec *instancetypev1beta1.VirtualMachinePreferenceSpec, vmiSpec *virtv1.VirtualMachineInstanceSpec) (*k8sfield.Path, error) {
+	if instancetypeSpec != nil && instancetypeSpec.Memory.Guest.Cmp(preferenceSpec.Requirements.Memory.Guest) < 0 {
```

Comment:
Should we check `Guest` for `nil`?

Reply:
I don't think `Guest` can be nil? It's a value, no pointer.